src/main/java/Duke.java

```diff
@@ -1,10 +1,134 @@
+                        duke.listTasks(duke);
+                        break;
+                }
+
+                if (action.equals(KeywordChecker.EXACT_KEYWORD_BYE)) {
+                    break;
+                }
+            } else if (KeywordChecker.containsMarkKeyword(action)) {
+                try {
+                    int id = KeywordChecker.getSpecifier(action);
+
+                    if (id > duke.tasks.size() || id < 1) {
+                        throw new DukeException("Out of bounds index!");
+                    }
+
+                    switch (KeywordChecker.getNonexactKeyword(action)) {
+                        case KeywordChecker.MARK_KEYWORD_MARK:
+                            duke.markTask(duke, id);
+                            break;
+                        case KeywordChecker.MARK_KEYWORD_UNMARK:
+                            duke.unmarkTask(duke, id);
+                            break;
+                        case KeywordChecker.MARK_KEYWORD_DELETE:
+                            duke.deleteTask(duke, id);
+                            break;
+                    }
+                } catch (DukeException ex) {
+                    duke.echo("Index is out of bounds!");
+                }
+            } else if (KeywordChecker.containsTaskKeyword(action)) {
+                try {
+                    switch (KeywordChecker.getNonexactKeyword(action)) {
+                        case KeywordChecker.TASK_KEYWORD_TODO:
+                            duke.tasks.add(new Todo(action));
+                            break;
+                        case KeywordChecker.TASK_KEYWORD_DEADLINE:
+                            if (!action.contains(Deadline.DELIMITER)) {
+                                throw new DukeException("Missing Deadline delimiter!");
+                            }
+                            duke.tasks.add(new Deadline(action));
+                            break;
+                        case KeywordChecker.TASK_KEYWORD_EVENT:
+                            if (!action.contains(Event.DELIMITER)) {
+                                throw new DukeException("Missing Event delimiter!");
+                            }
+                            duke.tasks.add(new Event(action));
+                            break;
+                    }
+
+                    duke.echo("Added task:" +
+                            duke.tasks.getLast().printTask());
+                } catch (DukeException ex) {
+                    duke.echo("Incorrect use of " + KeywordChecker.getNonexactKeyword(action));
+                }
+            } else {
+                duke.echo("I can't understand. Please enter a proper command, friend!");
+            }
+        }
+    }
+
+    public void listTasks(Duke interaction) {
+        interaction.tasks.forEach(
+                task -> { System.out.println(String.format("%d.%s",
+                        tasks.indexOf(task) + 1,
+                        task.printTask()));
+                });
```

Comment:
Maybe this function could be optimised to make use of lambda functions.
The function `printf` also provides the functionality of both `println` and `String.format`.

```
interaction.tasks.forEach(
    task -> System.out.printf("%d.%s%n",
        tasks.indexOf(task) + 1,
        task.printTask()));
```